File: Bwl.Framework/src/Settings/Gui/SettingField.vb
Public Class SettingField
    Private WithEvents setting As SettingOnStorage
    Private settingReady As Boolean
    Private _designText As String
    Sub New()
        InitializeComponent()
    End Sub
    Public Property AssignedSetting() As SettingOnStorage
        Get
            Return setting
        End Get
        Set(value As SettingOnStorage)
            setting = value
            ShowFields()
        End Set
    End Property


    Private Sub ShowFields()
		settingReady = False
		tbValue.PasswordChar = Nothing
        If setting Is Nothing Then
            Me.lCaption.Text = "Нет связанной настройки!"
            Me.lDesc.Text = "Нет связанной настройки!"
            cValue.Hide()
            cbValue.Hide()
            tbValue.Show()
            tbValue.Enabled = False
        Else
            If setting.FriendlyName.Length > 0 Then
                Me.lCaption.Text = setting.FriendlyName + " (" + setting.Name + ")"
            Else
                Me.lCaption.Text = setting.Name
            End If
            Me.lDesc.Text = setting.Description
            cValue.Hide()
            cbValue.Hide()
            tbValue.Hide()
			tbValue.Enabled = True
			lKey.Hide()
			tbKey.Hide()
            If TypeOf setting Is BooleanSetting Then
                Dim tmp = DirectCast(setting, BooleanSetting)
                cValue.Checked = tmp.Value
                cValue.Text = setting.Name
                cValue.Show()
            ElseIf TypeOf setting Is VariantSetting Then
                Dim tmp = DirectCast(setting, VariantSetting)
                cbValue.Items.Clear()
                For Each item In tmp.Variants
                    cbValue.Items.Add(item)
                Next
                cbValue.Text = tmp.Value
				cbValue.Show()
			ElseIf TypeOf setting Is PasswordSetting Then
				Dim passSet = CType(setting, PasswordSetting)
				tbValue.Text = passSet.Pass
				tbValue.PasswordChar = "*"c
				tbValue.Show()
				If (passSet.Key IsNot Nothing) Then
					tbKey.Text = String.Join(",", passSet.Key)
				Else
					settingReady = True
					tbKey_TextChanged(Nothing, Nothing)
					settingReady = False
				End If
				'tbKey.PasswordChar = "*"c
				tbKey.Show()
				lKey.Show()
			Else
				tbValue.Text = setting.ValueAsString
				tbValue.Show()
			End If

			If TypeOf setting Is StringSetting Then
				menuFile.Visible = True
			Else
				menuFile.Visible = False
			End If

		End If
		settingReady = True
	End Sub
	Public Overloads Sub Refresh() Handles setting.ParametersChanged, setting.ValueChanged
		If settingReady Then
			ShowFields()
			MyBase.Refresh()
		End If
	End Sub
	Private Sub SettingField_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load

	End Sub
	Private Sub cbValue_SelectionChangeCommitted(sender As System.Object, e As System.EventArgs) Handles cbValue.SelectionChangeCommitted

	End Sub
	Public Property DesignText() As String
		Get
			Return _designText
		End Get
		Set(value As String)
			_designText = value
			If setting Is Nothing Then
				lCaption.Text = _designText
			End If
		End Set
	End Property
	Public Event SettingValueChanged()
	Private Sub tbValue_TextChanged(sender As System.Object, e As System.EventArgs) Handles tbValue.TextChanged
		If settingReady And (Not Me.IsDisposed) Then
			settingReady = False
			If TypeOf setting Is PasswordSetting Then
				Dim tmp As PasswordSetting = DirectCast(setting, PasswordSetting)

				If tbValue.Text <> tmp.Pass Then
					Try
						tmp.Pass = tbValue.Text
						RaiseEvent SettingValueChanged()
					Catch ex As Exception
					End Try
				End If
			Else
				Try
					If tbValue.Text <> setting.ValueAsString Then
						setting.ValueAsString = tbValue.Text
						RaiseEvent SettingValueChanged()
					End If
				Catch ex As Exception
				End Try
			End If

			'  ShowFields()
			settingReady = True
		End If
	End Sub

	Private Sub cValue_Click(sender As Object, e As System.EventArgs) Handles cValue.Click
		If settingReady And (Not Me.IsDisposed) Then
			If TypeOf setting Is BooleanSetting Then
				DirectCast(setting, BooleanSetting).Value = cValue.Checked
				RaiseEvent SettingValueChanged()
			End If
		End If
		ShowFields()
	End Sub

	Private Sub cbValue_TextChanged(sender As Object, e As System.EventArgs) Handles cbValue.TextChanged
		If settingReady And (Not Me.IsDisposed) Then
			If TypeOf setting Is VariantSetting Then
				Dim tmp As VariantSetting = DirectCast(setting, VariantSetting)

				If cbValue.Text <> tmp.Value Then
					Try
						DirectCast(setting, VariantSetting).Value = cbValue.Text
						RaiseEvent SettingValueChanged()
					Catch ex As Exception
					End Try
				End If
				ShowFields()
			End If
		End If
	End Sub

	Private Sub menuDefault_Click(sender As System.Object, e As System.EventArgs)
		If settingReady And (Not Me.IsDisposed) Then
			settingReady = False
			setting.ValueAsString = setting.DefaultValueAsString
			ShowFields()
			settingReady = True
		End If
	End Sub

	Private Sub bMenu_LinkClicked(sender As System.Object, e As System.Windows.Forms.LinkLabelLinkClickedEventArgs) Handles bMenu.LinkClicked
		Dim text As String = setting.DefaultValueAsString.ToString
		menuDefault.Text = "По умолчанию (" + text + ")"
		menu.Show(MousePosition.X, MousePosition.Y)
	End Sub

	Private Sub menuFile_Click(sender As System.Object, e As System.EventArgs) Handles menuFile.Click
		If selectFile.ShowDialog = Windows.Forms.DialogResult.OK Then
			tbValue.Text = selectFile.FileName
		End If
	End Sub

	Private Sub tbKey_TextChanged(sender As Object, e As EventArgs) Handles tbKey.TextChanged
		If settingReady And (Not Me.IsDisposed) Then
			If TypeOf setting Is PasswordSetting Then
				Dim tmp As PasswordSetting = DirectCast(setting, PasswordSetting)
				Try
					tmp.Key = tbKey.Text.Split(","c).Select(Function(c) Convert.ToByte(c)).ToArray
					RaiseEvent SettingValueChanged()
				Catch ex As Exception
				End Try
				ShowFields()
			End If
		End If
	End Sub
End Class

File: Framework/Framework/Bwl.Framework.Windows/Settings/Gui/SettingField.cs
using System;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Runtime.Versioning;
using System.Windows.Forms;

namespace Bwl.Framework.Windows
{

    [SupportedOSPlatform("windows")]
    public partial class SettingField
    {

        private SettingOnStorage _setting;

        private SettingOnStorage setting
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return _setting;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_setting != null)
                {
                    _setting.ParametersChanged -= (_) => Refresh();
                    _setting.ValueChanged -= (_) => Refresh();
                }

                _setting = value;
                if (_setting != null)
                {
                    _setting.ParametersChanged += (_) => Refresh();
                    _setting.ValueChanged += (_) => Refresh();
                }
            }
        }
        private bool settingReady;
        private string _designText;

        public SettingField()
        {
            InitializeComponent();
        }

        public SettingOnStorage AssignedSetting
        {
            get
            {
                return setting;
            }
            set
            {
                setting = value;
                ShowFields();
            }
        }


        private void ShowFields()
        {
            settingReady = false;
            tbValue.PasswordChar = default;
            if (setting is null)
            {
                lCaption.Text = "Нет связанной настройки!";
                lDesc.Text = "Нет связанной настройки!";
                cValue.Hide();
                cbValue.Hide();
                tbValue.Show();
                tbValue.Enabled = false;
            }
            else
            {
                if (setting.FriendlyName.Length > 0)
                {
                    lCaption.Text = setting.FriendlyName + " (" + setting.Name + ")";
                }
                else
                {
                    lCaption.Text = setting.Name;
                }
                lDesc.Text = setting.Description;
                cValue.Hide();
                cbValue.Hide();
                tbValue.Hide();
                tbValue.Enabled = true;
                lKey.Hide();
                tbKey.Hide();

                SetBiggerField?.Invoke(false);
                tbValue.Multiline = false;
                tbValue.Height = 20;
                tbValue.ScrollBars = ScrollBars.None;

                if (setting is BooleanSetting)
                {
                    BooleanSetting tmp = (BooleanSetting)setting;
                    cValue.Checked = tmp.Value;
                    cValue.Text = setting.Name;
                    cValue.Show();
                }
                else if (setting is VariantSetting)
                {
                    VariantSetting tmp = (VariantSetting)setting;
                    cbValue.Items.Clear();
                    foreach (var item in tmp.GetVariants())
                        cbValue.Items.Add(item);
                    cbValue.Text = tmp.Value;
                    cbValue.Show();
                }
                else if (setting is PasswordSetting)
                {
                    PasswordSetting passSet = (PasswordSetting)setting;
                    SetBiggerField?.Invoke(true);
                    tbValue.Text = passSet.Password;
                    tbValue.PasswordChar = '*';
                    tbValue.Show();
                    if (passSet.Key is not null)
                    {
                        tbKey.Text = string.Join(",", passSet.Key);
                    }
                    else
                    {
                        settingReady = true;
                        tbKey_TextChanged(null, null);
                        settingReady = false;
                    }
                    tbKey.Show();
                    lKey.Show();
                    // Adjust the height of the control to accommodate the password field
                    Height = 150; // Adjust this value as needed
                }
                else if (setting is TextFileContentSetting)
                {
                    TextFileContentSetting settingValue = (TextFileContentSetting)setting;
                    SetBiggerField?.Invoke(true);
                    tbValue.Height = 85; // Default is 20
                    tbValue.ScrollBars = ScrollBars.Both;
                    tbValue.Multiline = true;
                    tbValue.Lines = settingValue.Value;
                    tbValue.Show();
                    tbValue.SelectionStart = tbValue.Text.Length;
                    tbValue.ScrollToCaret();
                }
                else
                {
                    tbValue.Text = setting.ValueAsString;
                    tbValue.Show();
                }

                if (setting.IsReadOnly)
                {
                    tbValue.ReadOnly = true;
                }

                menuFile.Visible = setting is StringSetting;

            }
            settingReady = true;
        }

        public new void Refresh()
        {
            if (settingReady)
            {
                ShowFields();
                base.Refresh();
            }
        }

        private void SettingField_Load(object sender, EventArgs e)
        {

        }

        private void cbValue_SelectionChangeCommitted(object sender, EventArgs e)
        {

        }

        public string DesignText
        {
            get
            {
                return _designText;
            }
            set
            {
                _designText = value;
                if (setting is null)
                {
                    lCaption.Text = _designText;
                }
            }
        }

        public event SettingValueChangedEventHandler SettingValueChanged;

        public delegate void SettingValueChangedEventHandler();
        public event SetBiggerFieldEventHandler SetBiggerField;

        public delegate void SetBiggerFieldEventHandler(bool value);

        private void tbValue_TextChanged(object sender, EventArgs e)
        {
            if (settingReady && !IsDisposed)
            {
                settingReady = false;
                if (setting is PasswordSetting)
                {
                    PasswordSetting tmp = (PasswordSetting)setting;

                    if ((tbValue.Text ?? "") != (tmp.Password ?? ""))
                    {
                        try
                        {
                            tmp.Password = tbValue.Text;
                            SettingValueChanged?.Invoke();
                        }
                        catch (Exception ex)
                        {
                        }
                    }
                }
                else if (setting is TextFileContentSetting)
                {
                    try
                    {
                        TextFileContentSetting origSetting = (TextFileContentSetting)setting;
                        origSetting.Value = tbValue.Lines;
                        SettingValueChanged?.Invoke();
                    }
                    catch (Exception ex)
                    {
                    }
                }
                else
                {
                    try
                    {
                        if ((tbValue.Text ?? "") != (setting.ValueAsString ?? ""))
                        {
                            setting.ValueAsString = tbValue.Text;
                            SettingValueChanged?.Invoke();
                        }
                    }
                    catch (Exception ex)
                    {
                    }
                }

                // ShowFields()
                settingReady = true;
            }
        }

        private void cValue_Click(object sender, EventArgs e)
        {
            if (settingReady && !IsDisposed)
            {
                if (setting is BooleanSetting)
                {
                    ((BooleanSetting)setting).Value = cValue.Checked;
                    SettingValueChanged?.Invoke();
                }
            }
            ShowFields();
        }

        private void cbValue_TextChanged(object sender, EventArgs e)
        {
            if (settingReady && !IsDisposed)
            {
                if (setting is VariantSetting)
                {
                    VariantSetting tmp = (VariantSetting)setting;

                    if ((cbValue.Text ?? "") != (tmp.Value ?? ""))
                    {
                        try
                        {
                            ((VariantSetting)setting).Value = cbValue.Text;
                            SettingValueChanged?.Invoke();
                        }
                        catch (Exception ex)
                        {
                        }
                    }
                    ShowFields();
                }
            }
        }

        private void menuDefault_Click(object sender, EventArgs e)
        {
            if (settingReady && !IsDisposed)
            {
                settingReady = false;
                setting.ValueAsString = setting.DefaultValueAsString;
                ShowFields();
                settingReady = true;
            }
        }

        private void bMenu_LinkClicked(object sender, LinkLabelLinkClickedEventArgs e)
        {
            string text = setting.DefaultValueAsString.ToString();
            menuDefault.Text = "По умолчанию (" + text + ")";
            menu.Show(MousePosition.X, MousePosition.Y);
        }

        private void menuFile_Click(object sender, EventArgs e)
        {
            if (selectFile.ShowDialog() == DialogResult.OK)
            {
                tbValue.Text = selectFile.FileName;
            }
        }

        private void tbKey_TextChanged(object sender, EventArgs e)
        {
            if (settingReady && !IsDisposed)
            {
                if (setting is PasswordSetting)
                {
                    PasswordSetting tmp = (PasswordSetting)setting;
                    try
                    {
                        tmp.Key = tbKey.Text.Split(',').Select(c => Convert.ToByte(c)).ToArray();
                        SettingValueChanged?.Invoke();
                    }
                    catch (Exception ex)
                    {
                    }
                    ShowFields();
                }
            }
        }

        private void Refresh(Setting setting) => Refresh();

    }
}